### ISSUE TITLE
Issues with hardware accelerated streaming?

### ISSUE THREAD

**Issue by a-dekker** [heartfin/harbour-sailfin#56]
Just installed a Jellyfin docker container for the first time and tried Sailfin before activating hardware acceleration on the server. Sailfin seems to play okay, but streaming cost of a lot of CPU power on the server side.

After enabling hardware acceleration the CPU load is gone, but I can no longer play videos using Sailfin. Debug information says `Media status: 7` on the screen. It plays fine on my kodi device using the Jellyfin plugin, in various browsers on SFOS and my desktop with this hardware acceleration enabled.

Any idea what is going wrong?

This is a large part of the server logging:
```
[22:13:35] [INF] [17] MediaBrowser.MediaEncoding.Transcoding.TranscodeManager: /usr/lib/jellyfin-ffmpeg/ffmpeg -analyzeduration 200M -probesize 1G -init_hw_device vaapi=va:,kernel_driver=i915,driver=iHD -init_hw_device qsv=qs@va -filter_hw_device qs -
hwaccel vaapi -hwaccel_output_format vaapi -noautorotate -i file:"/media/MyMovie.1080p.WEB.h264.mkv" -noautoscale -map_metadata -1 -map_chapters -1 -threads 0 -map 0:0 -map 0:1 -map -0:s -codec:v:0 h264_qsv -low
_power 1 -preset veryfast -look_ahead 0 -b:v 4872000 -maxrate 4872000 -bufsize 9744000 -profile:v:0 baseline -level 41 -g:v:0 75 -keyint_min:v:0 75 -vf "setparams=color_primaries=bt709:color_trc=bt709:colorspace=bt709,scale_vaapi=format=nv12:extra_hw_
frames=24,hwmap=derive_device=qsv,format=qsv" -codec:a:0 copy -copyts -avoid_negative_ts disabled -max_muxing_queue_size 2048 -f hls -max_delay 5000000 -hls_time 3 -hls_segment_type mpegts -start_number 0 -hls_segment_filename "/cache/transcodes/611eb
28456eb46a90096b2aa6c1e0a4e%d.ts" -hls_playlist_type vod -hls_list_size 0 -y "/cache/transcodes/611eb28456eb46a90096b2aa6c1e0a4e.m3u8"
[22:13:35] [ERR] [17] MediaBrowser.MediaEncoding.Transcoding.TranscodeManager: FFmpeg exited with code 1
[22:13:35] [ERR] [17] Jellyfin.Api.Middleware.ExceptionMiddleware: Error processing request. URL GET /videos/999a6d03-c38f-1352-db97-0130b3b6e3f0/hls1/main/0.ts.
MediaBrowser.Common.FfmpegException: FFmpeg exited with code 1
   at MediaBrowser.MediaEncoding.Transcoding.TranscodeManager.StartFfMpeg(StreamState state, String outputPath, String commandLineArguments, Guid userId, TranscodingJobType transcodingJobType, CancellationTokenSource cancellationTokenSource, String wo
rkingDirectory)
   at Jellyfin.Api.Controllers.DynamicHlsController.GetDynamicSegment(StreamingRequestDto streamingRequest, Int32 segmentId)
   at Jellyfin.Api.Controllers.DynamicHlsController.GetHlsVideoSegment(Guid itemId, String playlistId, Int32 segmentId, String container, Int64 runtimeTicks, Int64 actualSegmentLengthTicks, Nullable`1 static, String params, String tag, String devicePr
ofileId, String playSessionId, String segmentContainer, Nullable`1 segmentLength, Nullable`1 minSegments, String mediaSourceId, String deviceId, String audioCodec, Nullable`1 enableAutoStreamCopy, Nullable`1 allowVideoStreamCopy, Nullable`1 allowAudio
StreamCopy, Nullable`1 breakOnNonKeyFrames, Nullable`1 audioSampleRate, Nullable`1 maxAudioBitDepth, Nullable`1 audioBitRate, Nullable`1 audioChannels, Nullable`1 maxAudioChannels, String profile, String level, Nullable`1 framerate, Nullable`1 maxFram
erate, Nullable`1 copyTimestamps, Nullable`1 startTimeTicks, Nullable`1 width, Nullable`1 height, Nullable`1 maxWidth, Nullable`1 maxHeight, Nullable`1 videoBitRate, Nullable`1 subtitleStreamIndex, Nullable`1 subtitleMethod, Nullable`1 maxRefFrames, N
ullable`1 maxVideoBitDepth, Nullable`1 requireAvc, Nullable`1 deInterlace, Nullable`1 requireNonAnamorphic, Nullable`1 transcodingMaxAudioChannels, Nullable`1 cpuCoreLimit, String liveStreamId, Nullable`1 enableMpegtsM2TsMode, String videoCodec, Strin
g subtitleCodec, String transcodeReasons, Nullable`1 audioStreamIndex, Nullable`1 videoStreamIndex, Nullable`1 context, Dictionary`2 streamOptions)
   at lambda_method880(Closure, Object)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ActionMethodExecutor.TaskOfActionResultExecutor.Execute(ActionContext actionContext, IActionResultTypeMapper mapper, ObjectMethodExecutor executor, Object controller, Object[] arguments)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.<InvokeActionMethodAsync>g__Awaited|12_0(ControllerActionInvoker invoker, ValueTask`1 actionResultValueTask)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.<InvokeNextActionFilterAsync>g__Awaited|10_0(ControllerActionInvoker invoker, Task lastTask, State next, Scope scope, Object state, Boolean isCompleted)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.Rethrow(ActionExecutedContextSealed context)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.Next(State& next, Scope& scope, Object& state, Boolean& isCompleted)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.<InvokeInnerFilterAsync>g__Awaited|13_0(ControllerActionInvoker invoker, Task lastTask, State next, Scope scope, Object state, Boolean isCompleted)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeNextResourceFilter>g__Awaited|25_0(ResourceInvoker invoker, Task lastTask, State next, Scope scope, Object state, Boolean isCompleted)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.Rethrow(ResourceExecutedContextSealed context)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.Next(State& next, Scope& scope, Object& state, Boolean& isCompleted)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeFilterPipelineAsync>g__Awaited|20_0(ResourceInvoker invoker, Task lastTask, State next, Scope scope, Object state, Boolean isCompleted)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeAsync>g__Awaited|17_0(ResourceInvoker invoker, Task task, IDisposable scope)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeAsync>g__Awaited|17_0(ResourceInvoker invoker, Task task, IDisposable scope)
   at Jellyfin.Api.Middleware.ServerStartupMessageMiddleware.Invoke(HttpContext httpContext, IServerApplicationHost serverApplicationHost, ILocalizationManager localizationManager)
   at Jellyfin.Api.Middleware.WebSocketHandlerMiddleware.Invoke(HttpContext httpContext, IWebSocketManager webSocketManager)
   at Jellyfin.Api.Middleware.IPBasedAccessValidationMiddleware.Invoke(HttpContext httpContext, INetworkManager networkManager)
   at Jellyfin.Api.Middleware.LanFilteringMiddleware.Invoke(HttpContext httpContext, INetworkManager networkManager, IServerConfigurationManager serverConfigurationManager)
   at Microsoft.AspNetCore.Authorization.AuthorizationMiddleware.Invoke(HttpContext context)
   at Jellyfin.Api.Middleware.QueryStringDecodingMiddleware.Invoke(HttpContext httpContext)
   at Swashbuckle.AspNetCore.ReDoc.ReDocMiddleware.Invoke(HttpContext httpContext)
   at Swashbuckle.AspNetCore.SwaggerUI.SwaggerUIMiddleware.Invoke(HttpContext httpContext)
```

**Comment by HenkKalkwater** [comment]
Hello,
could you also send the FFMPEG log for the media file by chance? The standard log has not enough information. You should be able to find it in the Dashboard under Logs of the Jellyfin Web interface. 

Detecting when to transcode and when to directly stream files is still a part of Sailfin that is underdeveloped. You can always uncheck the _Allow transcoding_ option in the Sailfin settings under *Streaming settings* if it causes problems.

**Comment by a-dekker** [comment]
The ffmeg logging always ends with `No VA display found for any default device`.
[FFMPEG-log.zip](https://github.com/user-attachments/files/15915016/FFMPEG-log.zip)

Disabling "Allow transcoding" does not work either, it does not even try to start this way it seems.

**Comment by HenkKalkwater** [comment]
Since you mentioned running it in Docker, have you followed the steps required for hardware acceleration to work within Docker[^1]? Since the logs mention it cannot find the GPU, you may have forgotten to expose it to the Docker container.

[^1]: https://jellyfin.org/docs/general/administration/hardware-acceleration/intel#configure-with-linux-virtualization

**Comment by a-dekker** [comment]
I was under the impression I did not need to pass any devices to the container, as I saw the CPU load drop to a few percentage after a few minutes. But `vainfo` indeed tells me hw-acceleration is not activated inside the container. After changing the docker config, HW acceleration does get detected. And Sailfin now also starts working, so only Sailfin was affected by my misconfiguration. CPU is now also significantly lower (75% instead of 380%) when streaming to Sailfin. Perhaps this can still be tweaked, will have to test.

So you set me on the right track. Thanks Henk, will close this issue.
